pl-faded-parsons.py

```diff
@@ -1,314 +1,538 @@
+        data: dict # TODO: expand this, the "tag" tells us the type of JSON object this is
+
+    main: Trays
+    log: List[LogEntry] = field(default_factory=list)
+
+@dataclass(frozen=True)
+class Mustache:
+
+    @dataclass(frozen=True)
+    class Line:
+
+        @dataclass(frozen=True)
+        class Segment:
+            @dataclass
+            class Blank:
+                default: str
+                width: int
+
+            @dataclass
+            class Code:
+                content: str
+                language: str
+
+            # { 'blank': { 'default': ..., 'width': max(4, len(...) + 1) } }
+            blank: Union[Blank, None] = None
+            # { 'code': { 'content': ..., 'language': ... } }
+            code: Union[Code, None] = None
+
+        indent: int
+        segments: List[Segment]
+
+    @dataclass(frozen=True)
+    class TrayLines:
+        lines: List["Mustache.Line"] # [Line.to_mustache(l, lang) for l in lines]
+        size: bool = True # any truthy value will do
+
+    @dataclass(frozen=True)
+    class PrePostText:
+        text: str
+        language: str
+
+    # chevron skips rendering when values are falsy (eg pre-text/post-text/starter)
+
+    # main element config
+    answers_name : str
+    language : str
+    previous_log : str
+    uuid: str
+
+    # trays and code context
+    starter: Union[TrayLines, Literal['']]
+    pre_text: PrePostText
+    given: TrayLines
+    post_text: PrePostText
 
-        parsed_segments = [{ "code" : { "content" : segments[0] } }]
-        for segment, pre_fill in zip(segments[1:], fills):
-            width = str(len(pre_fill)+1) if pre_fill != "" else "4"
-            parsed_segments.append({ "blank" : { "default" : pre_fill, "width" : width } })
-            parsed_segments.append({ "code"  : { "content" : segment  } })
-
-        matches = re.search(r'#([0-9]+)given', tail)
-        if matches is not None:
-            indent = int(matches.group(1))
-            new_line['indent'] = indent * indent_size
-            parsed_segments[-1] = {
-                "code" : { "content" : re.sub(r'#([0-9]+)given', '', tail).rstrip() }
-            }
-            new_line['segments'] = parsed_segments
-            given.append(new_line)
-            continue
-
-        if re.match(r'#distractor', tail):
-            parsed_segments[-1] = {
-                "code" : { "content" : re.sub(r'#distractor', '', tail).rstrip() }
-            }
-            new_line['segments'] = parsed_segments
-            distractors.append(new_line)
-            continue
-
-        new_line['segments'] = parsed_segments
-        scrambled.append(new_line)
-
-
-    for _ in range(max_distractors):
-        if len(distractors) == 0:
-            break
-        index = random.randint(len(distractors))
-        scrambled.append(distractors.pop(index))
-
-    return scrambled, given
+#
+# Helper Routines
+#
 
+def get_child_text_by_tag(element, tag: str) -> str:
```

Comment:
Can we make this an inner function somewhere? I think it's only used in one spot...